extensions/amp-gwd-animation/0.1/amp-gwd-animation-impl.js

```diff
@@ -0,0 +1,447 @@
+   * animations are re-enabled and the CSS override is lifted.
+   * @param {boolean} enable True to enable, false to disable.
+   */
+  setEnabled(enable) {
+    if (enable) {
+      // Lift the animation CSS override.
+      this.ampdoc_.getBody().classList.remove(ANIMATIONS_DISABLED_CLASS);
+    } else {
+      // Stop all animations by overriding them with `animation: none`.
+      this.ampdoc_.getBody().classList.add(ANIMATIONS_DISABLED_CLASS);
+    }
+  }
+
+  /**
+   * Stops animations on the previously-active page and enables them on the
+   * newly-active page.
+   * @param {number} index The index of the newly-active slide.
+   */
+  setCurrentPage(index) {
+    // Turn off animations on the previously-active page, if there was one.
+    // TODO(sklobovskaya): Decide if it's worth just storing the index.
+    const currentPageEl = this.ampdoc_.getRootNode().querySelector(
+        `.${GWD_PAGE_WRAPPER_CLASS}.${PlaybackCssClass.PLAY}`);
+
+    if (currentPageEl) {
+      currentPageEl.classList.remove(PlaybackCssClass.PLAY);
+    }
+
+    // Activate animations on the new current page.
+    const gwdPages = this.ampdoc_.getRootNode().querySelectorAll(
+        `.${GWD_PAGE_WRAPPER_CLASS}`);
+    const newPageEl = gwdPages[index];
+
+    if (newPageEl) {
+      newPageEl.classList.add(PlaybackCssClass.PLAY);
+    }
+  }
+
+  /**
+   * The play action.
+   * @param {string} id Receiver id.
+   */
+  play(id) {
+    const receiver = this.getReceiver(id);
+
+    if (!receiver) {
+      return;
+    }
+
+    receiver.classList.remove(PlaybackCssClass.PAUSE);
+  }
+
+  /**
+   * The pause action.
+   * @param {string} id Receiver id.
+   */
+  pause(id) {
+    const receiver = this.getReceiver(id);
+
+    if (!receiver) {
+      return;
+    }
+
+    receiver.classList.add(PlaybackCssClass.PAUSE);
+  }
+
+  /**
+   * The togglePlay action.
+   * @param {string} id Receiver id.
+   */
+  togglePlay(id) {
+    const receiver = this.getReceiver(id);
+
+    if (!receiver) {
+      return;
+    }
+
+    receiver.classList.toggle(PlaybackCssClass.PAUSE);
+  }
+
+  /**
+   * The gotoAndPlay action.
+   * @param {string} id Receiver id.
+   * @param {string} label The name of the label animation to go to.
+   */
+  gotoAndPlay(id, label) {
+    const receiver = this.getReceiver(id);
+
+    if (!receiver) {
+      return;
+    }
+
+    this.playLabelAnimation_(receiver, label);
+  }
+
+  /**
+   * The gotoAndPause action. A gotoAndPause is a gotoAndPlay followed by an
+   * (almost) immediate pause.
+   * @param {string} id Receiver id.
+   * @param {string} label The name of the label animation to go to.
+   */
+  gotoAndPause(id, label) {
+    const receiver = this.getReceiver(id);
+
+    if (!receiver) {
+      return;
+    }
+
+    // Switch to the label animation.
+    this.playLabelAnimation_(receiver, label);
+
+    // Trigger a pause.
+    const timeToWaitBeforePause = 40;
```

Comment:
Why 40ms specifically?

Could this be a module-level constant?

Reply:
Yes, this should be a constant. Done.

The 40ms delay is necessary to work around a still-present Safari bug (in our regular non-AMP runtime, too). I added some comments in the code to mention this.